lib/optimizely/odp/odp_event_manager.rb

```diff
@@ -0,0 +1,249 @@
+      logger: NoOpLogger.new,
+      proxy_config: nil
+    )
+      super()
+      @odp_config = odp_config
+      @mutex = Mutex.new
+      @event_queue = SizedQueue.new(Optimizely::Helpers::Constants::ODP_EVENT_MANAGER[:DEFAULT_QUEUE_CAPACITY])
+      @queue_capacity = Helpers::Constants::ODP_EVENT_MANAGER[:DEFAULT_QUEUE_CAPACITY]
+      # received signal should be sent after adding item to event_queue
+      @received = ConditionVariable.new
+      @logger = logger
+      @zaius_manager = api_manager || ZaiusRestApiManager.new(logger: @logger, proxy_config: proxy_config)
+      @batch_size = Helpers::Constants::ODP_EVENT_MANAGER[:DEFAULT_BATCH_SIZE]
+      @flush_interval = Helpers::Constants::ODP_EVENT_MANAGER[:DEFAULT_FLUSH_INTERVAL]
+      @flush_deadline = 0
+      @retry_count = Helpers::Constants::ODP_EVENT_MANAGER[:DEFAULT_RETRY_COUNT]
+      # current_batch should only be accessed by processing thread
+      @current_batch = []
+      @thread_exception = false
+    end
+
+    def start!
+      if running?
+        @logger.log(Logger::WARN, 'Service already started.')
+        return
+      end
+      @thread = Thread.new { run }
+      @logger.log(Logger::INFO, 'Starting scheduler.')
+    end
+
+    def flush
+      begin
+        @event_queue.push(:FLUSH_SIGNAL, non_block: true)
+      rescue ThreadError
+        @logger.log(Logger::ERROR, 'Error flushing ODP event queue.')
+        return
+      end
+
+      @mutex.synchronize do
+        @received.signal
+      end
+    end
+
+    def dispatch(event)
+      if @thread_exception
+        @logger.log(Logger::ERROR, format(Helpers::Constants::ODP_LOGS[:ODP_EVENT_FAILED], 'Queue is down'))
+        return
+      end
+
+      # if the processor has been explicitly stopped. Don't accept tasks
+      unless running?
+        @logger.log(Logger::WARN, 'ODP event queue is shutdown, not accepting events.')
+        return
+      end
+
+      begin
+        @logger.log(Logger::DEBUG, 'ODP event queue: adding event.')
+        @event_queue.push(event, non_block: true)
+      rescue => e
+        @logger.log(Logger::WARN, format(Helpers::Constants::ODP_LOGS[:ODP_EVENT_FAILED], e.message))
+        return
+      end
+
+      @mutex.synchronize do
+        @received.signal
+      end
+    end
+
+    def send_event(type:, action:, identifiers:, data:)
+      case @odp_config.odp_state
+      when OdpConfig::ODP_CONFIG_STATE[:UNDETERMINED]
+        @logger.log(Logger::DEBUG, 'ODP event queue: cannot send before the datafile has loaded.')
+        return
+      when OdpConfig::ODP_CONFIG_STATE[:NOT_INTEGRATED]
+        @logger.log(Logger::DEBUG, Helpers::Constants::ODP_LOGS[:ODP_NOT_INTEGRATED])
+        return
+      end
+
+      event = Optimizely::OdpEvent.new(type: type, action: action, identifiers: identifiers, data: data)
+      dispatch(event)
+    end
+
+    def stop!
+      return unless running?
+
+      begin
+        @event_queue.push(:SHUTDOWN_SIGNAL, non_block: true)
```

Comment:
The events that are received after this `SHUTDOWN` signal should be dropped. It appears that the event rejection depends on `isRunning` which will not become false unless `SHUTDOWN_SIGNAL` has been processed and the thread is done with. If my understanding is correct, it might be helpful to add a boolean `shutdown_started` or something along those lines which can be immediately set to `true` as soon as `stop` is called and the event manager should start rejecting events as soon as `shutdown_started` becomes true. What do you think?

Reply:
SizedQueue has a close method that seems designed for this exact scenario- seems like a good solution?

Reply:
Looks pretty neat to me 👍